### PR TITLE
Lifecycle and version

### PR DESCRIPTION
**Pull request by Zstro**

### CODE REVIEW

File: api/datatype/lifecycle/utils.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package lifecycle
+
+import . "github.com/journeymidnight/yig/error"
+
+type StorageClass uint8
+
+// Reference：https://docs.aws.amazon.com/zh_cn/AmazonS3/latest/dev/storage-class-intro.html
```
Comment (cannium) [inline api/datatype/lifecycle/utils.go]: Why define them again?

Comment (Zstro) [reply api/datatype/lifecycle/utils.go]: because of import cycle

File: .travis.yml [hunk 2]
```diff
@@ -31,12 +31,12 @@ script:
   - make integrate
   # golang 1.12 will change the permission of go.mod and go.sum
   # use chmod to reset the permission of go.mod and go.sum to make go test happy
-
   - sudo chmod 644 go.mod go.sum
   - sudo python test/sanity.py
   - pushd test/go
   - go test -v
   - popd
+  - cat lc.log
```
[the remark this answers is not in the document]
Comment (Zstro) [reply .travis.yml]: OK

File: api/bucket-lifecycle-handlers.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+func (api ObjectAPIHandlers) GetBucketLifeCycleHandler(w http.ResponseWriter, r *http.Request) {
+	reqCtx := GetRequestContext(r)
+	logger := reqCtx.Logger
+
+	var credential common.Credential
+	var err error
+	switch signature.GetRequestAuthType(r) {
+	default:
```
[the remark this answers is not in the document]
Comment (Zstro) [reply api/bucket-lifecycle-handlers.go]: OK

File: conf/yig.toml [hunk 4]
```diff
@@ -18,6 +18,12 @@ enable_pprof = false
 pprof_listener = "0.0.0.0:8730"
 reserved_origins = "s3.test.com,s3-internal.test.com"
 
+# LC
+lc_thread = 64
+# 10s
+lifecycle_spec = "*/20 * * * * ?"
```
[the remark this answers is not in the document]
Comment (fireworkmarks) [reply conf/yig.toml]: like @every 20m

[the remark this answers is not in the document]
Comment (fireworkmarks) [reply conf/yig.toml]: or @every 20s

File: api/datatype/lifecycle/utils.go (hunk 1 above)
Comment (cannium) [reply api/datatype/lifecycle/utils.go]: Eliminate it.

File: api/access-log-handler.go [hunk 5]
```diff
@@ -97,4 +97,4 @@ func NewAccessLogHandler(handler http.Handler, _ *meta.Meta) http.Handler {
 		handler: handler,
 		format:  format,
 	}
-}
+}
```
[the remark this answers is not in the document]
Comment (cxt90730) [reply api/access-log-handler.go]: cmd `go-fmt`

File: meta/types/storage-class.go [hunk 6]
```diff
@@ -11,23 +11,23 @@ const (
 	// ObjectStorageClassStandard is a ObjectStorageClass enum value
 	ObjectStorageClassStandard StorageClass = iota
 
-	// ObjectStorageClassReducedRedundancy is a ObjectStorageClass enum value
-	ObjectStorageClassReducedRedundancy
```
Comment (cxt90730) [inline meta/types/storage-class.go]: Not need to change

File: test/go/lib/client.go [hunk 7]
```diff
@@ -23,6 +23,7 @@ const (
 	TEST_KEY            = "testput"
 	TEST_KEY_SPECIAL    = "testputspecial:!@$%^&*()_+=-;?><| "
 	TEST_VALUE          = "valueput"
+	TEST_STORAGEGLACIER = "GLACIER "
```
Comment (cxt90730) [inline test/go/lib/client.go]: use const ; and the spelling of value is fault

File: meta/lifecycle.go [hunk 8]
```diff
@@ -10,6 +10,10 @@ func LifeCycleFromBucket(b Bucket) (lc LifeCycle) {
 
 func (m *Meta) PutBucketToLifeCycle(bucket Bucket) error {
 	lifeCycle := LifeCycleFromBucket(bucket)
+	err := m.Client.RemoveBucketFromLifeCycle(bucket)
```
Comment (cxt90730) [inline meta/lifecycle.go]: it should be a transaction?

File: meta/types/storage-class.go (hunk 6 above)
Comment (Zstro) [reply meta/types/storage-class.go]: I use enumeration order when resolving conflicts；e.x：GLACIER > STANDARD_IA > STANDARD

File: meta/lifecycle.go (hunk 8 above)
Comment (Zstro) [reply meta/lifecycle.go]: Otherwise every time putLifecycle is inserted

File: conf/yig.toml [hunk 9]
```diff
@@ -18,6 +18,14 @@ enable_pprof = false
 pprof_listener = "0.0.0.0:8730"
 reserved_origins = "s3.test.com,s3-internal.test.com"
 
+# LC
+lc_thread = 64
+# LC Debug
+lc_debug = true
+# 5s for pass CI
+lifecycle_spec = "@every 5s"
```
Comment (cxt90730) [inline conf/yig.toml:26]: it's also a debug parameter

File: conf/yig.toml [hunk 10]
```diff
@@ -18,6 +18,14 @@ enable_pprof = false
 pprof_listener = "0.0.0.0:8730"
 reserved_origins = "s3.test.com,s3-internal.test.com"
 
+# LC
+lc_thread = 64
+# LC Debug
+lc_debug = true
```
Comment (cxt90730) [inline conf/yig.toml:24]: I suggest 'debug mode' is used for global env，but not only lifecycle

File: api/bucket-lifecycle-handlers.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+			WriteErrorResponse(w, r, err)
+			return
+		}
+	default:
+		// For all unknown auth types return error.
+		WriteErrorResponse(w, r, ErrAccessDenied)
+		return
+	}
+
+	lifecycle, err := api.ObjectAPI.GetBucketLifecycle(reqCtx, credential)
```
Comment (cxt90730) [inline api/bucket-lifecycle-handlers.go:64]: To be sure the situation of `ErrSqlNoRows`

File: meta/object.go [hunk 12]
```diff
@@ -118,6 +118,11 @@ func (m *Meta) UpdateGlacierObject(reqCtx RequestContext, targetObject, sourceOb
 		if err != nil {
 			return err
 		}
+	} else if onlyTranStorageClass {
```
Comment (cxt90730) [inline meta/object.go:121]: TODO：check

File: test/go/lifecycle_test.go [hunk 13]
```diff
@@ -1,127 +1,245 @@
 package _go
 
-//import (
```
Comment (cxt90730) [inline test/go/lifecycle_test.go:3]: remove unused code

File: tools/lc.go [hunk 14]
```diff
@@ -215,17 +416,17 @@ func processLifecycle() {
...
 			if err != nil {
-				helper.Logger.Error("Bucket", item.BucketName, "retrieve error:", err)
+				helper.Logger.Error("Bucket", item.BucketName, "Lifecycle process error:", err)
 				waitgroup.Done()
 				continue
 			}
 			helper.Logger.Info("Bucket lifecycle done:", item.BucketName)
 		default:
-			if empty == true {
+			if wait == true {
```
Comment (cannium) [inline tools/lc.go:428]: Clear the meaning of variable "wait"

File: meta/common/storage-class.go [hunk 15]
```diff
@@ -50,6 +48,16 @@ var (
...
 	}
+
+	StorageClassWeight = map[StorageClass]int{
+		ObjectStorageClassStandard:           1,
+		ObjectStorageClassStandardIa:         2,
+		ObjectStorageClassIntelligentTiering: 3,
+		ObjectStorageClassOnezoneIa:          4,
+		ObjectStorageClassGlacier:            5,
+		ObjectStorageClassDeepArchive:        6,
+		ObjectStorageClassReducedRedundancy:  1,
```
[the remark this answers is not in the document]
Comment (Zstro) [reply meta/common/storage-class.go:59]: Sorting for resolve StorageClass transition conflicts.

**Pull request closed without merging**